src/main/java/com/slack/eithernet/Retries.kt

```diff
@@ -0,0 +1,83 @@
+ *
+ * @param maxAttempts The maximum number of times to retry the operation. Default is 5.
+ * @param initialDelay The delay before the first retry. Default is 1 second.
+ * @param delayFactor The factor by which the delay should increase after each failed attempt.
+ *   Default is 2.0.
+ * @param maxDelay The maximum delay between retries. Default is [Duration.INFINITE].
+ * @param jitterFactor The maximum factor of jitter to introduce. For example, a value of 0.1 will
+ *   introduce up to 10% jitter. Default is 0.
+ * @param onFailure An optional callback for failures, useful for logging.
+ * @param block The block of code to retry. This block should return an [ApiResult].
+ * @return The result of the operation if it's successful, or the last failure result if all
+ *   attempts fail.
+ *
+ * This function will attempt the operation you give it up to [maxAttempts] times, multiplying the
+ * delay between each attempt by [delayFactor], starting from [initialDelay] and not exceeding
+ * [maxDelay]. If the operation continues to fail after [maxAttempts] times, it will return the last
+ * failure result. If the operation succeeds at any point, it will immediately return the success
+ * result.
+ *
+ * Note: This uses a default exponential backoff strategy with optional jitter. Depending on your
+ * use case, you might want to customize the strategy, for example by handling certain kinds of
+ * failures differently.
+ */
+@Suppress("LongParameterList", "ReturnCount")
+public suspend fun <T : Any, E : Any> retryWithExponentialBackoff(
+  maxAttempts: Int = 5,
+  initialDelay: Duration = 1.seconds,
+  delayFactor: Double = 2.0,
+  maxDelay: Duration = Duration.INFINITE,
+  jitterFactor: Double = 0.0,
+  onFailure: ((attempt: Int, result: ApiResult.Failure<E>) -> Unit)? = null,
+  block: suspend () -> ApiResult<T, E>
+): ApiResult<T, E> {
+  require(maxAttempts > 0) { "maxAttempts must be greater than 0" }
+  var currentDelay = initialDelay
+  repeat(maxAttempts) { attempt ->
+    when (val result = block()) {
+      is ApiResult.Success -> return result
+      is ApiResult.Failure -> {
+        onFailure?.invoke(attempt, result)
+        if (attempt == maxAttempts - 1) {
+          return result // return last failure
+        } else {
+          delay(currentDelay)
+          // Compute a new delay using a combination of the factor and optional jitter.
+          currentDelay = (currentDelay * delayFactor).coerceAtMost(maxDelay)
+          if (jitterFactor != 0.0) {
+            // Note that Random.nextDouble requires a range,
+            // so we provide it with -jitterFactor to jitterFactor.
+            // It also cannot be 0, so only do this if jitter is non-zero
+            val jitter = 1 + Random.nextDouble(-jitterFactor, jitterFactor)
+            currentDelay = (currentDelay * jitter).coerceAtMost(maxDelay)
+          }
+        }
+      }
+    }
+  }
+  return ApiResult.unknownFailure(RuntimeException("Max attempts reached"))
```

Comment:
Unreachable?

Reply:
Reachable!

Reply:
ehh not reachable, but compiler doesn't trust it [6471ecb](https://github.com/slackhq/EitherNet/pull/60/commits/6471ecb39a6e64d87bd38ad7f776310599f7365c)